text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+
+# Rust language interface and interaction with other features
+[language-interface]: #language-interface
+
+The Rust implementation of crABI will include:
+- An ABI for defining functions or referencing external functions:
+  `extern "crabi"`.
+- A repr for laying out data structures (`struct`, `union`, `enum`) compatible
+  with crABI: `repr(crabi)`.
+- A new lint `improper_crabi_types`, analogous to `improper_ctypes`. This lint
+  defaults to `deny`.
+
+crABI structures and functions also support any structure defined using
+`repr(C)`, as well as enums that have a discriminant type specified.
+
+crABI structures and functions support types defined using `repr(transparent)`
+if they would support the type of the field.
+
+For an enum, the crABI `repr` allows additionally specifying the discriminant
+type: `repr(crabi, u8)`. If an `enum` specifies `repr(crabi)` but does not
+specify a discriminant type, the enum is guaranteed to use the smallest
+discriminant type that holds the maximum discriminant value used by a variant
+in the enum. (This differs from the behavior of `repr(C)` enums without a
+discriminant type.)
+
+crABI supports passing and storing pointers to `repr(Rust)` structures, to
+allow using these opaque pointers as handles, but does not support passing such
+structures by value.
+
+crABI works in combination with `repr(align)`, with the same meaning.
+
+crABI does not support `repr(packed)`.
+
+# crABI versioning and evolution
+[crabi-versioning-and-evolution]: #crabi-versioning-and-evolution
+
+crABI has has a major and minor version number, similar to semver.
+
+The major version number will change if crABI changes *incompatibly*; for
+instance, crABI 2.0 would not necessarily interoperate with crABI 1.3.
+
+The minor version number will change if crABI changes *compatibly*, such that
+newer callers can understand and work with older functions and structures.
+Older callers cannot necessarily understand and work with newer functions and
+structures, unless the newer side restricts itself to features understood by an
+older version.
+
+This RFC defines crABI 1.0.
+
+Versions of crABI will start out as unstable features, and follow the usual
+Rust stabilization process. As with any other nightly feature, versions of
+crABI may, prior to stabilization, change in ways incompatible with prior
+*nightly-only* implementations of that version of crABI. Versions of crABI
+should not be considered stable until available in stable Rust.
+
+Future versions of crABI may also establish allow-by-default lints for the use
+of features newer than a particular crABI version; such lints will allow
+libraries or modules exposing an ABI to restrict themselves to versions of
+crABI compatible with specific other software (e.g. languages or runtimes) they
+want to interoperate with.
+
+An implementation of crABI should document which version of crABI it
+implements, which compactly conveys supported and unsupported functionality.
+
```

Comment:
Having a global `__crabi_version` would cause conflicts if multiple static libraries that export a crABI interface are linked together. And using COMDAT to deduplicatw wouldn't work either if both static libraries use a different crABI version.